src/pkcs11/pkcs11-global.c

```diff
@@ -282,8 +296,12 @@ CK_RV C_Initialize(CK_VOID_PTR pInitArgs)
 	in_finalize = 0;
```

Comment:
The above block already changes global context by calling C_Finalize() or changing initialized_pid. Should not lock come even before this block? Rather safe than sorry.

Reply:
I don't think so. 
The DeleteCriticalSection is only used on windows, and the module_close is called from DLLMain when the process is unloading the DLL.  the C_Finalize is called the application before the application unloads a module, as PKCS11 does not define how modules are loaded/linked. 

The code is designed to have a static mutex already setup when C_Initialize is called to protect the code that sets up the current locking for pkcs11, either: none, OS locking or what ever the calling application provides. This static mutex is is not the PKCS11 global_lock. (I am not even sure if the call has to be made when the DLL is being deleted.) 

The pthread version is  also in static storage, and initialized via PTHREAD_MUTEX_INITIALIZER which is done by the OS and pthread when linking or loading, so no extra code is needed.

Reply:
OR are you saying  line 304  `C_INITIALIZE_M_LOCK` should come before line 300 `in_finalize = 0;` ?